src/iarc7_sensors/orientation_filter.py

```diff
@@ -0,0 +1,184 @@
+
+        Returns `-1` if there are no messages of type `klass` in the
+        queue
+        '''
+        for i in xrange(len(self._queue)-1, -1, -1):
+            if (type(self._queue[i][1]) == klass
+                 and (time is None or self._queue[i][1].header.stamp < time)):
+                return i
+        return -1
+
+    def _callback(self, klass, msg):
+        with self._lock:
+            last_i = self._get_last_index(klass)
+            if msg.header.stamp <= self._queue[last_i][1].header.stamp:
+                # This message isn't newer than the last one of type klass
+                rospy.logwarn(
+                    ('Ignoring message of type {} with timestamp {} older than '
+                   + 'previous message with timestamp {}')
+                        .format(
+                            klass,
+                            msg.header.stamp,
+                            self._queue[last_i][1].header.stamp))
+                return
+
+            for i in xrange(len(self._queue) - 1, -1, -1):
+                if self._queue[i][1].header.stamp < msg.header.stamp:
+                    index = i+1
+                    break
+            else:
+                index = 0
+
+            data = [float('NaN')] * 3
+            try:
+                data[1] = -msg.data.pitch
+                data[2] = msg.data.roll
+            except AttributeError:
+                # If this message type doesn't have pitch and roll, skip it
+                pass
+
+            self._queue.insert(index, (data, msg))
+            self._reprocess(index)
+
+            if (not math.isnan(self._queue[-1][0][0])
+                and not math.isnan(self._queue[-1][0][1])
+                and not math.isnan(self._queue[-1][0][2])):
+
+                new_stamp = max(self._last_published_stamp + rospy.Duration(0, 1),
+                                self._queue[-1][1].header.stamp)
+                self._publish_transform(*reversed(self._queue[-1][0]), time=new_stamp)
+                self._last_published_stamp = new_stamp
+
+            while len(self._queue) > self._message_queue_length:
+                self._queue.pop(0)
+
+    def _reprocess(self, index):
+        for i in xrange(index, len(self._queue)):
+            last_yaw = self._queue[i-1][0][0]
+            msg = self._queue[i][1]
+            last_orientation_i = self._get_last_index(OrientationAnglesStamped,
+                                                      msg.header.stamp)
+
+            if type(msg) == OrientationAnglesStamped:
+
+                if last_orientation_i == -1:
+                    dyaw = 0
+                else:
+                    dyaw = -(msg.data.yaw
+                           - self._queue[last_orientation_i][1].data.yaw)
+
+                if dyaw > math.pi:
+                    dyaw -= 2*math.pi
+                elif dyaw < -math.pi:
+                    dyaw += 2*math.pi
+
+                self._queue[i][0][0] = last_yaw + dyaw
+            else:
```

Comment:
Might be a battle of style here, but I would prefer elif type(msg) == Float64Stamped and an else clause that throws an exception of some kind or logs a warning